### PR TITLE
`ap`, applicative "idiom brackets" sugar

### PR DESCRIPTION
**Pull request by blendmaster**
Essentially the "idiom brackets" from the [original Applicative paper][0] as well as the [syntax proposal for Haskell][1] that show up from time to time. Unfortunately (or fortunately, depending on who you ask),
clojure doesn't have reader macros, so we can't implement `(| |)` banana brackets.

It's actually hard to think up a good name for this that isn't silly or overly long. `ap` seems pretty clear in the clojure world, but suggestions are welcome. 

[0]: http://staff.city.ac.uk/~ross/papers/Applicative.pdf
[1]: https://ocharles.org.uk/IdiomBrackets.html

### CODE REVIEW

**Comment by niwinz** [comment]
Nice work! I like the naming, so `ap` is good enough for me.
Would be awesome have tests for it before merge.

**Comment by yurrriq** [comment]
:+1: for this with tests!

**Comment by blendmaster** [comment]
I dwelt on the name some more and I'm still not entirely happy with it; It seems too short to fit in with other clojure.core stuff like `constantly`, `identity` (vs `const` and `id`). The one precedent it has is [Haskell's `ap`][0], but that's replaced by `<$>` and `<*>` in real usage.

Some straw man proposals:

- `applyap`: apply with applicative effects. a bit wordy, but less mysterious.
- `app` / `appl` / `applic`: basically the same descriptive power as  `ap`,
   but might lend themselves to a cuter explanation.
- `$*`: the two Haskell symbols, smashed together. Sort of follows the precent of `->` and
  core.async's `>!` `<!` for all-symbol macro names, to disambiguate from a regular
  function. Slightly harder to type than `ap` though. Alternatively, `<$*>`.
-  `<$>` (or `<*>`), i.e. break backwards compatibility with the existing aliases for fmap/fapply
  and repurpose as this macro.

[0]: http://hackage.haskell.org/package/base-4.8.1.0/docs/Control-Monad.html#v:ap

Besides the name, I've got a couple other questions:

- Can I get nitpicks on the docstring? I want to make sure it fits with the rest of the docstrings in this library, as well as actually being useful to somebody who didn't write the macro itself.
- Speaking of docstrings, is there a stable port of python's doctest to clojure? It'd be cool
to actually test the examples in there. I found https://github.com/drojas/doctest , but it's not finished.
- Opinions on creating a new `cats.labs` namespace for this, and potentially #121 too? Maybe `cats.labs.sugar` or something; Mainly don't want to add anything to the core namespace without vetting usefulness.
- The implementation of the macro literally uses `alet` right now; it is somewhat nice for explanatory power, but given the complexity of the `alet` macro, I wonder if it's better to expand `(ap f aa bb)` directly to `(fapply (fn [a] (fn [b] (f a b)) aa bb)`; all the dependency checking in `alet` is unnecessary here.

**Comment by yurrriq** [comment]
#### Name
- I dislike `$*` and `<$*>`.
- `applyap` seems to follow the `constantly` and `identity` trend nicely.
- I don't like `applic`. And `appl` seems too easy to mistype or confuse for `#'clojure.core/apply`.
- `app` seems like an ok choice. (But then again, why not just `ap` at that point?)

#### Docstring
I'll try to give it a thorough reading tomorrow and report back once I have.

#### doctest
Looks cool, but I don't know of anything like that.

#### `cats.labs.sugar` idea
:+1:

#### Macro expansion
Maybe instead of expanding to `(fapply ...)` we could add an option to `alet` (or create an `alet*`) that elides dependency checking. N.B. I haven't looked at `alet` for a while.

**Comment by niwinz** [comment]
My two cents:

Naming: I agree with @yurrriq, I don't like the haskell like cryptic names. And prefer something simple. +1 to `ap`.

About docstrings: I'm not clerarly understand you :(
About doctest: I don't like them. Tests should be live separate to the code I think.

I like the idea of put that under labs and if we think that is usefull, merge it into core. So +1 a it.

And about macro expansion: I think that using alet is not a problem. The macro expansions is done at compile time so it does not have any performance implications in runtime. But review you the tradeoffs and choice the appropriate approach for you. Later we can revisit that if something works unexpectedly ;)

File: src/cats/core.cljc [hunk 1]
```diff
@@ -534,6 +534,39 @@
...
+
+   If you're familiar with Haskell, this is closest to writing
+   \"in Applicative style\": you can straightforwardly convert
+   pure function application to effectful application by with
+   some light syntax (<$> and <*> in case of Haskell, and `ap` here).
+
+   See the original Applicative paper for more inspiration:
+   http://staff.city.ac.uk/~ross/papers/Applicative.pdf"
+  [f & args]
+  (let [syms (for [_ args] (gensym))]
```
Comment (dialelo) [inline src/cats/core.cljc]: we've used the `(repeatedly (count args) gensym)` idiom in other parts of the code, maybe we should change it here too?

**Comment by dialelo** [comment]
Great addition to the library, thanks for your time @blendmaster! I think that `cats.labs.sugar` is a good idea for testing out syntactic sugar ideas but I'd leave `ap` in `cats.core` since it's such a handy operator.

**Comment by blendmaster** [comment]
Moved to `cats.labs.sugar`, added tests. While the functionality is definitely core-worthy, I still want more feedback on the name. Once people get used to how code with `ap` looks and don't complain, then we can move it into core.

Also, I added a family of `ap`-style threading macros, because I could. They _might_ be useful, but I dunno.

**Comment by dialelo** [comment]
>Moved to cats.labs.sugar, added tests. While the functionality is definitely core-worthy, I still want more feedback on the name. Once people get used to how code with ap looks and don't complain, then we can move it into core.

Excellent, thanks a lot for your time.

>Also, I added a family of ap-style threading macros, because I could. They might be useful, but I dunno.

Let's see if they are, I like them!

**Comment by dialelo** [comment]
I guess is not WIP anymore, right?

**Comment by blendmaster** [comment]
Yep, complete enough for me.

**Comment by dialelo** [comment]
Thanks a lot @blendmaster, `ap` is my new favorite cats macro!

**Comment by niwinz** [comment]
nice work!

But, now a little bit of documentation would be awesome! :P
